File: python-backend/backend-python/db/replicationManager.py
# backend-python/db/replicationManager.py
from core.dbManager import db
from core.eventBus import eventBus
from utils.logger import logger

async def replicateEdgeToShared():
    """
    Replicate all edge DB collections to shared DB
    """
    logger.log("[ReplicationManager] Starting edge → shared replication...")
    for collection, items in db.store.get("edge", {}).items():
        for key, value in items.items():
            await db.set(collection, key, value, target="shared")
            eventBus.publish("db:update", {"collection": collection, "key": key, "value": value, "source": "replicationManager"})
    logger.log("[ReplicationManager] Edge → shared replication complete.")

# Subscribe to edge updates for live replication
async def on_edge_update(event):
    await db.set(event["collection"], event["key"], event["value"], target="shared")

eventBus.subscribe("db:update", on_edge_update)

File: typescript-backend/lib/db/replicationManager.ts
// src/db/replicationManager.ts
import { db } from "./dbManager";
import { eventBus } from "../core/eventBus";
import { logger } from "../utils/logger";

/**
 * Replicate all records from edge → shared store
 * Can be called periodically or triggered by db:update
 */
export async function replicateEdgeToShared(collection?: string) {
  const collections = collection ? [collection] : Object.keys(db["edge"]);
  for (const col of collections) {
    const records = await db.getAll(col, "edge");
    for (const record of records) {
      await db.set(col, record.id, record, "shared");
      logger.log(`[Replication] ${col}:${record.id} replicated to shared`);
    }
  }
}

// Listen for all db:update events and replicate automatically
eventBus.subscribe("db:update", async ({ collection }) => {
  if (!collection) return;
  await replicateEdgeToShared(collection);
});
